src/frequenz/sdk/timeseries/resampling.py

```diff
@@ -0,0 +1,405 @@
+
+        Note that the error could be raised by the sink, while trying to send
+        a resampled data for this timeseries, the source key is only used to
+        identify the timeseries with the issue, it doesn't necessarily mean
+        that the error was raised by the source. The underlying exception
+        should provide information about what was the actual source of the
+        exception.
+        """
+
+    def __repr__(self) -> str:
+        """Return the representation of the instance.
+
+        Returns:
+            The representation of the instance.
+        """
+        return f"{self.__class__.__name__}({self.exceptions=})"
+
+
+class Resampler:
+    """A timeseries resampler.
+
+    In general timeseries [`Source`][frequenz.sdk.timeseries.Source]s don't
+    necessarily come at periodic intervals. You can use this class to normalize
+    timeseries to produce `Sample`s at regular periodic intervals.
+
+    This class uses
+    a [`ResamplingFunction`][frequenz.sdk.timeseries.ResamplingFunction] to
+    produce a new sample from samples received in the past. If there are no
+    samples coming to a resampled timeseries for a while, eventually the
+    `Resampler` will produce `Sample`s with `None` as value, meaning there is
+    no way to produce meaningful samples with the available data.
+    """
+
+    def __init__(
+        self,
+        *,
+        resampling_period_s: float,
+        resampling_function: ResamplingFunction,
+        max_data_age_in_periods: float = 3.0,
+    ) -> None:
+        """Initialize an instance.
+
+        Args:
+            resampling_period_s: The time it passes between resampled data
+                should be calculated (in seconds).
+            max_data_age_in_periods: The maximum age a sample can have to be
+                considered *relevant* for resampling purposes, expressed in the
+                number of resampling periods. For exapmle is
+                `resampling_period_s` is 3 and `max_data_age_in_periods` is 2,
+                then data older than `3*2 = 6` secods will be discarded when
+                creating a new sample and never passed to the resampling
+                function.
+            resampling_function: The function to be applied to the sequence of
+                *relevant* samples at a given time. The result of the function
+                is what is sent as the resampled data.
+        """
+        self._resampling_period_s = resampling_period_s
+        self._max_data_age_in_periods: float = max_data_age_in_periods
+        self._resampling_function: ResamplingFunction = resampling_function
+        self._resamplers: Dict[Source, _StreamingHelper] = {}
+        self._timer: Timer = Timer(self._resampling_period_s)
+
+    async def stop(self) -> None:
+        """Cancel all receiving tasks."""
+        await asyncio.gather(*[helper.stop() for helper in self._resamplers.values()])
+
+    def add_timeseries(self, source: Source, sink: Sink) -> bool:
```

Comment:
I'm not sure if providing any support for callbacks is a good idea, especially in interfaces that will be exposed to the users.  Maybe one of these options is better:

1.  `add_timeseries` returns a `MetricResampler` instance, which provides an `__anext__` method,  that will produce a value whenever a call to `Resampler.resample()` finishes.
2. `Resampler.resample()` returns a list of tuples each containing a name for a stream, and the latest `Sample`.

Reply:
I was thinking about going for 2, but make it an `AsyncIterator` producing the tuple (so `AsyncIterator[Tuple[Source, Sample]]`, so users can do:

```py
async for source, sample in resampler.resample():
    do_whatever(source, sample)
```

What do you think?

Reply:
> I was thinking about going for 2, but make it an AsyncIterator[Map[Source, Sample]], so users can do

also possible with:

``` python
t1 = resampler.add_timeseries(...)
t2 = resampler.add_timeseries(...)

while await resampler.resample():
    t1.consume()
```

and there would be no wrapping up of values in a map just to be unwrapped in the next step.  looks weirdly similar to some other parts of the code as well.  :D

But just putting it out there.  Maybe maps are better,  especially in dynamic contexts,  else people will have to have a second loop to go through the `MetricResampler` objects.  Although that will be the same performance as the `async for` approach,  it will be a few more lines of code.

Reply:
This would mean the `resample` method would no longer be long running I guess?

Reply:
Yes,  it is just that we are trying to pack so much into a single process already,  that we might end up having to prioritize performance over having a familiar interface.

But I agree,  hashing of small objects is very fast,  and would be insignificant compared to the cost of the surrounding `await`.  So we can go with this as well.

Reply:
I'm trying to apply this approach but exception handling gets more convoluted. If I go for the async iterator, then one approach is for it to be an `AsyncIterator[Dict[Souce, Union[Sample, Exception]]]` but then the user needs to check every resampling result to see if there was an error or not, while in the current approach if everything goes well you don't have to do anything and if only some go wrong, then what went well was delivered and you only get "notifications" of what went bad.

This issue applies too to your approach, which I'm also considering. The user will have to do a try/except for every `t.consume()` to see if it failed or not, which can litter the code quite a bit.

I keep thinking about how to do this in a way that it is convenient to the user, but throwing the problem out there in case you have any suggestions.

Reply:
For me that, besides being too unpythonic, is also not the whole story, because if there is an error with, for example `t1`, you probably won't want consume it, so the code will be still littered with `if` to check for errors and how to handle them.

I think I will make the `ResamplingError` to take also `samples`, so if there is a problem, you get the exceptions but also the samples that were produced correctly. It still sucks a bit as it makes error handling more complicated, but I guess if you are dealing with a lot of streams of data there is no real way around it.

Reply:
To add more context, I guess this approach makes sense because if you are resampling a group of metrics, you probably care about having a complete window, as long as that happens, you operate normally. If there is a problem, you need to inspect the situation, so you can get from the exception which metrics were calculated successfully and which didn't, and you can decide what to do based on that (abort the whole thing, use the ones that are available, etc.).

Reply:
> for example t1, you probably won't want consume it, so the code will be still littered with if to check for errors and how to handle them.

maybe you missed the `continue` statement in the example?  So `consume` would only be called if `resample` worked for all streams.

Btw,  channels closing could cause the resampler to fail,  what else could cause an exception in the resampler?

Reply:
Ah, yeah, I missed it. But then, if you don't consume when there is an error, you lost the whole window of data, the resampler will wait until the next timer fire if you continue.

And also I assume even if you don't call `consume()` then the samples of the current window should be discarded, otherwise different timeseries will get out of sync, or at least (if you remove the faulty ones) will be delayed, as now the new tick will send the samples from the old tick.

For example, the resampling actor will want to keep sending all the successful timeseries and remove the failing ones, not ignore the whole resampling window.

Still not so simple IMHO. But again, I'm more and more convinced that the complexity is inherent to the problem, there is no way to really take it away.

> Btw, channels closing could cause the resampler to fail, what else could cause an exception in the resampler?

The resampling function I guess (hello callbacks :D), not counting bugs is the only thing I can think of.

Reply:
> And also I assume even if you don't call consume() then the samples of the current window should be discarded, otherwise different timeseries will get out of sync.

Yes,  if some values are not consumed,  they will get overwritten by the next run.  I think that's fine.  And we could add a boolean check,  and log a warning saying that values are getting discarded.  People trying to start a resampler and not using the output is too basic a problem that we can't solve with any approach.

> For example, the resampling actor will want to keep sending all the successful timeseries and remove the failing ones, not ignore the whole resampling window.

If that happens,  we should not send any values out,  because if we do,  downstream code will go out of sync.  The formula engine can be made to resynchronize,  but it does that by dropping some values until all streams have caught up to the latest values.  I think it is better to do that in the resampler instead.